### ISSUE TITLE
Expected Training Speed?

### ISSUE THREAD

**Issue by Tomiha** [pierluigiferrari/ssd_keras#344]
First of all, I really appreciate your work and notebooks about this SSD implementation!
I started experimenting with a network that only detects faces, so your SSD with only 2 classes (I adjusted the pipeline according to the instruction in the weight_sampling_tutorial.

At the moment I'm training the network, but training speed is a bit low (I think). So I was wondering if others could give me some estimation on their training speeds.

Network I'm training:
SSD300 with adjustments to output 2 classes (background / face)
* 1323 Training Samples (data generator)
* 331 Validation Samples (data generator)
* Adam Optimizer
* batch_size 16 (ran into memory issues)
* 83 steps per epoch (training)

>> Runs at +-10min per epoch

Is this slow/normal/fast?

**Comment by fader111** [comment]
Hi, I use 1070 GPU, original training parameters from ssd_training.ipynb  - 1000 samples/epoch, 
batch -  32, here is it the process log:
1000/1000 [==============================] - 1843s 2s/step - loss: 3.8849 - val_loss: 3.2442

Epoch 00001: val_loss improved from 5.00000 to 3.24423, saving model to h5/trained/ssd300_pascal_07+12_epoch-01_loss-3.8844_val_loss-3.2442.h5
Epoch 2/120

Epoch 00002: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1898s 2s/step - loss: 3.8920 - val_loss: 3.2299

Epoch 00002: val_loss improved from 3.24423 to 3.22992, saving model to h5/trained/ssd300_pascal_07+12_epoch-02_loss-3.8929_val_loss-3.2299.h5
Epoch 3/120

Epoch 00003: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1893s 2s/step - loss: 3.8767 - val_loss: 3.2799

Epoch 00003: val_loss did not improve from 3.22992
Epoch 4/120

Epoch 00004: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1765s 2s/step - loss: 3.8613 - val_loss: 3.2833

Epoch 00004: val_loss did not improve from 3.22992
Epoch 5/120

Epoch 00005: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1649s 2s/step - loss: 3.8420 - val_loss: 3.2737

Epoch 00005: val_loss did not improve from 3.22992
Epoch 6/120

Epoch 00006: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1545s 2s/step - loss: 3.8250 - val_loss: 3.3087

Epoch 00006: val_loss did not improve from 3.22992
Epoch 7/120

Epoch 00007: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1496s 1s/step - loss: 3.8032 - val_loss: 3.2559

Epoch 00007: val_loss did not improve from 3.22992
Epoch 8/120

Epoch 00008: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1426s 1s/step - loss: 3.7966 - val_loss: 3.2717

Epoch 00008: val_loss did not improve from 3.22992
Epoch 9/120

Epoch 00009: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1419s 1s/step - loss: 3.7676 - val_loss: 3.2929

Epoch 00009: val_loss did not improve from 3.22992
Epoch 10/120

Epoch 00010: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1437s 1s/step - loss: 3.7666 - val_loss: 3.2666

Epoch 00010: val_loss did not improve from 3.22992
Epoch 11/120

Epoch 00011: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1395s 1s/step - loss: 3.7641 - val_loss: 3.2828

Epoch 00011: val_loss did not improve from 3.22992
Epoch 12/120

Epoch 00012: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1474s 1s/step - loss: 3.7415 - val_loss: 3.3097

Epoch 00012: val_loss did not improve from 3.22992
Epoch 13/120

Epoch 00013: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1569s 2s/step - loss: 3.7431 - val_loss: 3.3565

Epoch 00013: val_loss did not improve from 3.22992
Epoch 14/120

Epoch 00014: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1876s 2s/step - loss: 3.7292 - val_loss: 3.3566

Epoch 00014: val_loss did not improve from 3.22992
Epoch 15/120

Epoch 00015: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1748s 2s/step - loss: 3.7188 - val_loss: 3.3086

Epoch 00015: val_loss did not improve from 3.22992
Epoch 16/120

Epoch 00016: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1693s 2s/step - loss: 3.7067 - val_loss: 3.3149

Epoch 00016: val_loss did not improve from 3.22992
Epoch 17/120

Epoch 00017: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1684s 2s/step - loss: 3.6957 - val_loss: 3.3598

Epoch 00017: val_loss did not improve from 3.22992
Epoch 18/120

Epoch 00018: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1669s 2s/step - loss: 3.6846 - val_loss: 3.3439

Epoch 00018: val_loss did not improve from 3.22992
Epoch 19/120

Epoch 00019: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1644s 2s/step - loss: 3.6728 - val_loss: 3.3395

Epoch 00019: val_loss did not improve from 3.22992
Epoch 20/120

Epoch 00020: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1888s 2s/step - loss: 3.6842 - val_loss: 3.3458

Epoch 00020: val_loss did not improve from 3.22992
Epoch 21/120

Epoch 00021: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1891s 2s/step - loss: 3.6536 - val_loss: 3.3814

Epoch 00021: val_loss did not improve from 3.22992
Epoch 22/120

Epoch 00022: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1790s 2s/step - loss: 3.6516 - val_loss: 3.4011

Epoch 00022: val_loss did not improve from 3.22992
Epoch 23/120

Epoch 00023: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1877s 2s/step - loss: 3.6317 - val_loss: 3.3075

Epoch 00023: val_loss did not improve from 3.22992
Epoch 24/120

Epoch 00024: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1814s 2s/step - loss: 3.6339 - val_loss: 3.4036

Epoch 00024: val_loss did not improve from 3.22992
Epoch 25/120

Epoch 00025: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1843s 2s/step - loss: 3.6248 - val_loss: 3.3721

Epoch 00025: val_loss did not improve from 3.22992
Epoch 26/120

Epoch 00026: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1873s 2s/step - loss: 3.6317 - val_loss: 3.3623

Epoch 00026: val_loss did not improve from 3.22992
Epoch 27/120

Epoch 00027: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1880s 2s/step - loss: 3.6136 - val_loss: 3.3704

Epoch 00027: val_loss did not improve from 3.22992
Epoch 28/120

Epoch 00028: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1854s 2s/step - loss: 3.6122 - val_loss: 3.3643

Epoch 00028: val_loss did not improve from 3.22992
Epoch 29/120

Epoch 00029: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1876s 2s/step - loss: 3.5985 - val_loss: 3.3522

Epoch 00029: val_loss did not improve from 3.22992
Epoch 30/120

Epoch 00030: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1900s 2s/step - loss: 3.6084 - val_loss: 3.3526

Epoch 00030: val_loss did not improve from 3.22992
Epoch 31/120

Epoch 00031: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1884s 2s/step - loss: 3.5842 - val_loss: 3.4307

Epoch 00031: val_loss did not improve from 3.22992
Epoch 32/120

Epoch 00032: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1878s 2s/step - loss: 3.5870 - val_loss: 3.3578

Epoch 00032: val_loss did not improve from 3.22992
Epoch 33/120

Epoch 00033: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1923s 2s/step - loss: 3.5767 - val_loss: 3.3922

Epoch 00033: val_loss did not improve from 3.22992
Epoch 34/120

Epoch 00034: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1926s 2s/step - loss: 3.5669 - val_loss: 3.3783

Epoch 00034: val_loss did not improve from 3.22992
Epoch 35/120

Epoch 00035: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1922s 2s/step - loss: 3.5612 - val_loss: 3.3888

Epoch 00035: val_loss did not improve from 3.22992
Epoch 36/120

Epoch 00036: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1895s 2s/step - loss: 3.5709 - val_loss: 3.4509

Epoch 00036: val_loss did not improve from 3.22992
Epoch 37/120

Epoch 00037: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1876s 2s/step - loss: 3.5406 - val_loss: 3.3959

Epoch 00037: val_loss did not improve from 3.22992
Epoch 38/120

Epoch 00038: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1910s 2s/step - loss: 3.5275 - val_loss: 3.4135

Epoch 00038: val_loss did not improve from 3.22992
Epoch 39/120

Epoch 00039: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1880s 2s/step - loss: 3.5251 - val_loss: 3.4358

Epoch 00039: val_loss did not improve from 3.22992
Epoch 40/120

Epoch 00040: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1867s 2s/step - loss: 3.5298 - val_loss: 3.5070

Epoch 00040: val_loss did not improve from 3.22992
Epoch 41/120

Epoch 00041: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1854s 2s/step - loss: 3.5387 - val_loss: 3.4251

Epoch 00041: val_loss did not improve from 3.22992
Epoch 42/120

Epoch 00042: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1796s 2s/step - loss: 3.5264 - val_loss: 3.4424

Epoch 00042: val_loss did not improve from 3.22992
Epoch 43/120

Epoch 00043: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1757s 2s/step - loss: 3.4965 - val_loss: 3.4869

Epoch 00043: val_loss did not improve from 3.22992
Epoch 44/120

Epoch 00044: LearningRateScheduler setting learning rate to 0.001.
1000/1000 [==============================] - 1848s 2s/step - loss: 3.5212 - val_loss: 3.4507

Epoch 00044: val_loss did not improve from 3.22992
Epoch 45/120

I spend 2 days already. But can't see any loss improvement. That' strange...

**Comment by Tomiha** [comment]
Thanks for sharing!
Same thing here :(. 
No immediate clue why it's not improving...

Below is my output, this is from a resumed training. Stopped improving after epoch 43, so it stayed the same for +-100 epochs (till epoch 150). 

> Epoch 00144: LearningRateScheduler setting learning rate to 1e-05.
> 42/42 [==============================] - 169s 4s/step - loss: 4.2822 - val_loss: 5.3433
> 
> Epoch 00144: val_loss did not improve from 5.32624
> Epoch 145/150
> 
> Epoch 00145: LearningRateScheduler setting learning rate to 1e-05.
> 42/42 [==============================] - 163s 4s/step - loss: 4.1405 - val_loss: 5.3474
> 
> Epoch 00145: val_loss did not improve from 5.32624
> Epoch 146/150
> 
> Epoch 00146: LearningRateScheduler setting learning rate to 1e-05.
> 42/42 [==============================] - 165s 4s/step - loss: 3.9866 - val_loss: 5.3496
> 
> Epoch 00146: val_loss did not improve from 5.32624
> Epoch 147/150
> 
> Epoch 00147: LearningRateScheduler setting learning rate to 1e-05.
> 42/42 [==============================] - 164s 4s/step - loss: 4.0636 - val_loss: 5.3482
> 
> Epoch 00147: val_loss did not improve from 5.32624
> Epoch 148/150
> 
> Epoch 00148: LearningRateScheduler setting learning rate to 1e-05.
> 42/42 [==============================] - 168s 4s/step - loss: 4.0742 - val_loss: 5.3353
> 
> Epoch 00148: val_loss did not improve from 5.32624
> Epoch 149/150
> 
> Epoch 00149: LearningRateScheduler setting learning rate to 1e-05.
> 42/42 [==============================] - 154s 4s/step - loss: 4.0959 - val_loss: 5.3436
> 
> Epoch 00149: val_loss did not improve from 5.32624
> Epoch 150/150
> 
> Epoch 00150: LearningRateScheduler setting learning rate to 1e-05.
> 42/42 [==============================] - 164s 4s/step - loss: 4.1674 - val_loss: > 5.3368
>
> Epoch 00150: val_loss did not improve from 5.32624

Maybe I need to expand my dataset?
Or do we really have to use around 1000 steps per epoch?

**Comment by fader111** [comment]
> Maybe I need to expand my dataset?

I think It depends on the features of objects, you are looking for, but 42 examples seem very few number. All IMHO.

**Comment by Tomiha** [comment]
It's 42 steps per epoch, so it saw 42 * batch_size examples in total. But indeed network only updates 42 times per epoch. Again was a small test, I'm looking into Yolo at the moment.
